### PR TITLE
Implement error collection for deserialization (#1196)

### PR DESCRIPTION
**Pull request by sri-adarsh-kumar**
# PR: Implement error collection for deserialization (#1196)

## AI Disclosure

**This is an AI-generated solution.** The requirements were specified by a human developer based on a 9-year-old feature request from @odrotbohm, and the implementation was fully built by Claude Code (an AI coding agent).

## Overview

Adds opt-in error collection capability to Jackson deserialization, allowing applications to collect multiple deserialization errors instead of failing fast on the first error. This addresses issue #1196 requested by Oliver Drotbohm in 2016.

**Target branch**: `3.x` (for Jackson 3.1.0)

## Motivation

Currently, Jackson fails fast on the first deserialization error, requiring a fix-retry cycle to discover all problems in a payload. This is inefficient for:
- **API validation**: Return all validation errors to clients in one response
- **Batch processing**: Log all errors but continue processing
- **Development tooling**: Show developers all issues at once
- **Data migration**: Identify all problematic records upfront

### 1. Opt-in Per-Call Design
- **No global configuration** - must explicitly call `collectErrors()` on `ObjectReader`
- **Thread-safe** - each call gets its own isolated error bucket
- **Backward compatible** - default behavior unchanged

### 2. DoS Protection
- Default limit: 100 errors per call (configurable)
- Hard failure when limit exceeded, with collected errors as suppressed exceptions
- Prevents memory exhaustion from malicious payloads

### 3. Best-Effort Collection
- Collects recoverable errors (type mismatches, unknown properties, etc.)
- Non-recoverable errors (malformed JSON, structural issues) still fail immediately
- Failed fields get default values (0 for primitives, null for objects)

### 4. JSON Pointer Paths (RFC 6901)
- Each error includes path to problematic field: `/items/0/price`
- Proper escaping: `~` → `~0`, `/` → `~1`
- Makes it easy to locate issues in complex nested structures

### 5. Thread Safety
- Per-call error buckets - safe to reuse `ObjectReader` across threads
- No shared state between calls

### Basic Usage
```java
ObjectMapper mapper = new JsonMapper();
ObjectReader reader = mapper.readerFor(Order.class).collectErrors();

try {
    Order result = reader.readValueCollecting(json);
    // Success - all fields valid
} catch (DeferredBindingException ex) {
    System.out.println("Found " + ex.getProblems().size() + " problems:");
    for (CollectedProblem problem : ex.getProblems()) {
        System.out.println(problem.getPath() + ": " + problem.getMessage());
        // Can also access problem.getRawValue() to see bad input
    }
}
```

### Custom Error Limit
```java
// Collect up to 10 errors before hard failure
ObjectReader reader = mapper.readerFor(Order.class).collectErrors(10);
```

### Multiple Input Types
```java
// Supports all standard input types
Order order1 = reader.readValueCollecting(jsonString);
Order order2 = reader.readValueCollecting(jsonBytes);
Order order3 = reader.readValueCollecting(file);
Order order4 = reader.readValueCollecting(inputStream);
Order order5 = reader.readValueCollecting(reader);
```

#### 1. `CollectedProblem`
Immutable value object representing a single deserialization error:
- `path`: JSON Pointer to problematic field (e.g., `/items/0/price`)
- `message`: Human-readable error message
- `rawValue`: The invalid input value that caused the error
- `targetType`: Expected Java type

#### 2. `DeferredBindingException`
Aggregate exception thrown when errors are collected:
- `problems`: List of all collected errors
- `partialResult`: The partially deserialized object (with default values for failed fields)
- `limitReached`: Flag indicating if error limit was hit
- Extends `DatabindException` for compatibility

#### 3. `CollectingProblemHandler`
Stateless problem handler that collects errors during deserialization:
- Implements `DeserializationProblemHandler` interface
- No shared state - all state passed via `DeserializationContext`
- Handles type mismatches, unknown properties, and other recoverable errors
- Returns default values to allow deserialization to continue

#### 4. `ObjectReader` Extensions
New methods:
- `collectErrors()`: Enable collection with default limit (100)
- `collectErrors(int maxProblems)`: Enable collection with custom limit
- `readValueCollecting(...)`: Read with error collection (5 overloads for different input types)

### Design Decisions

1. **Per-call buckets**: Each `readValueCollecting()` call gets a fresh error list stored in `DeserializationContext` attributes
2. **Default values**: Failed fields use Java defaults (0, false, null) to allow deserialization to continue
3. **Suppressed exceptions**: When hard failure occurs (limit reached, non-recoverable error), collected problems are attached as suppressed exceptions
4. **No recursion**: Skips children of unknown properties to avoid cascading errors

## Testing

Comprehensive test suite with 31 tests covering:

### Core Functionality
- **Default fail-fast behavior** (2 tests): Unchanged when not using error collection
- **Per-call bucket isolation** (2 tests): Successive and concurrent calls don't share errors
- **Thread safety** (1 test): 10 threads processing in parallel with isolated buckets

### Edge Cases
- **JSON Pointer escaping** (4 tests): Tilde, slash, combined, array indices
- **Limit behavior** (3 tests): Default limit, custom limit, under limit
- **Unknown properties** (2 tests): Collection and child skipping
- **Type coercions** (6 tests): int, long, double, float, boolean, boxed types
- **Root-level errors** (2 tests): Non-recoverable, path formatting
- **Hard failures** (1 test): Suppressed exception attachment
- **Message formatting** (2 tests): Single error, multiple errors
- **Input types** (6 tests): byte[], File, InputStream, Reader, null handling, empty JSON

## Backward Compatibility

- **100% backward compatible** - no changes to existing behavior
- Feature is completely opt-in via explicit method calls
- New exception type extends existing `DatabindException`
- No performance impact when not using the feature
- No breaking changes to any existing APIs

## Performance Considerations

- **Negligible overhead when disabled**: Feature only activates when explicitly requested
- **Per-call allocation**: Each call allocates a small ArrayList for errors (default capacity: 10)
- **Early termination**: Stops collecting at configured limit to prevent DoS
- **No synchronization**: Thread-safe through isolation, not locking

### API Validation
```java
@PostMapping("/orders")
public ResponseEntity<?> createOrder(@RequestBody String json) {
    try {
        Order order = orderReader.readValueCollecting(json);
        return ResponseEntity.ok(orderService.save(order));
    } catch (DeferredBindingException ex) {
        List<ValidationError> errors = ex.getProblems().stream()
            .map(p -> new ValidationError(p.getPath().toString(), p.getMessage()))
            .toList();
        return ResponseEntity.badRequest().body(errors);
    }
}
```

### Batch Processing
```java
for (String record : records) {
    try {
        Data data = reader.readValueCollecting(record);
        process(data);
    } catch (DeferredBindingException ex) {
        logger.error("Record validation failed with {} errors:",
                     ex.getProblems().size());
        ex.getProblems().forEach(p ->
            logger.error("  {}: {}", p.getPath(), p.getMessage()));
        // Continue with next record
    }
}
```

## Limitations

1. **Best-effort**: Not all errors can be collected
   - Malformed JSON (missing braces) fails immediately
   - Structural problems fail immediately
   - Type conversion errors ARE collected
   - Unknown property errors ARE collected (if enabled)

2. **Default values**: Failed fields get defaults
   - Primitives: 0, false
   - Objects: null
   - May affect downstream validation logic

3. **Error limit**: Default 100 errors
   - Prevents DoS but may not catch all issues in very large payloads
   - Configurable per call

## Future Enhancements

Potential future improvements (not in this PR):
- Structured error types (beyond just messages)
- Integration with Bean Validation (JSR-380)
- Custom error formatters
- Error severity levels
- Configurable default value strategies

---

**Generated by**: Claude Code (Anthropic)
**Specified by**: @sri-adarsh-kumar
**Original feature request**: @odrotbohm (2016)

### CODE REVIEW

File: src/main/java/tools/jackson/databind/ObjectReader.java [hunk 1]
```diff
@@ -692,6 +692,61 @@ public ObjectReader withHandler(DeserializationProblemHandler h) {
...
+    /**
+     * Enables error collection mode with a custom problem limit.
+     *
+     * <p><b>Thread-safety</b>: The returned reader is immutable and thread-safe.
+     * Each call to {@link #readValueCollecting} allocates a fresh problem bucket,
+     * so concurrent calls do not interfere.
+     *
+     * @param maxProblems Maximum number of problems to collect (must be > 0)
+     * @return A new ObjectReader configured for error collection
+     * @since 3.1
```
Comment (pjfanning) [inline src/main/java/tools/jackson/databind/ObjectReader.java:740]: the throws should be included in the javadoc

File: src/main/java/tools/jackson/databind/ObjectReader.java [hunk 2]
```diff
@@ -1320,6 +1375,187 @@ public <T> T readValue(TokenBuffer src) throws JacksonException
...
+     * @return Deserialized object
+     * @throws tools.jackson.databind.exc.DeferredBindingException if recoverable problems were collected
+     * @throws tools.jackson.databind.DatabindException if a non-recoverable error occurred
+     * @since 3.1
+     */
+    public <T> T readValueCollecting(JsonParser p) throws JacksonException {
+        _assertNotNull("p", p);
+
+        // CRITICAL: Allocate a FRESH bucket for THIS call (thread-safety)
+        List<tools.jackson.databind.exc.CollectedProblem> bucket = new ArrayList<>();
```
Comment (pjfanning) [inline src/main/java/tools/jackson/databind/ObjectReader.java]: no need to use fully qualified class name here

File: src/main/java/tools/jackson/databind/ObjectReader.java [hunk 3]
```diff
@@ -1320,6 +1375,187 @@ public <T> T readValue(TokenBuffer src) throws JacksonException
...
+
+        // Create per-call attributes with the fresh bucket
+        ContextAttributes perCallAttrs = _config.getAttributes()
+            .withPerCallAttribute(tools.jackson.databind.deser.CollectingProblemHandler.class, bucket);
```
Comment (pjfanning) [inline src/main/java/tools/jackson/databind/ObjectReader.java]: again, avoid using fully qualified class names - use imports so the code can be kept concise

File: src/main/java/tools/jackson/databind/ObjectReader.java [hunk 4]
```diff
@@ -692,6 +692,61 @@ public ObjectReader withHandler(DeserializationProblemHandler h) {
...
+     *     .forType(MyBean.class)
+     *     .collectErrors();
+     *
+     * MyBean bean = reader.readValueCollecting(json);
+     * </pre>
+     *
+     * @return A new ObjectReader configured for error collection
+     * @since 3.1
+     */
+    public ObjectReader collectErrors() {
```
Comment (pjfanning) [inline src/main/java/tools/jackson/databind/ObjectReader.java]: I don't really like the names of the new APIs. I don't especially like 'collect' but I think 'Error' is plain wrong.
'Error' has a very explicit meaning in Java - examples include OutOfMemoryError. Exception is more correct than Error here.

**Comment by JooHyukKim** [comment]
-1 on this PR itself due to the sheer amount of line changes. 

Ai code assistants should really get trained on splitting tasks to safer(or at least feeling so) bits of progression

File: src/main/java/tools/jackson/databind/deser/CollectingProblemHandler.java [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+    }
+
+    /**
+     * Escapes a JSON Pointer segment per RFC 6901.
+     * Must escape '~' before '/' to avoid double-escaping.
+     *
+     * @param segment The raw segment (property name or array index)
+     * @return Escaped segment safe for JSON Pointer
+     */
+    private String escapeJsonPointerSegment(String segment) {
```
Comment (cowtowncoder) [inline src/main/java/tools/jackson/databind/deser/CollectingProblemHandler.java]: Should not duplicate code here instead of using `src/main/java/com/fasterxml/jackson/core/JsonPointer.java` (from `jackson-core`)

File: src/main/java/tools/jackson/databind/ObjectReader.java (hunk 4 above)
Comment (cowtowncoder) [reply src/main/java/tools/jackson/databind/ObjectReader.java]: Agreed.

Comment (cowtowncoder) [reply src/main/java/tools/jackson/databind/ObjectReader.java]: Something like `problemCollectingReader()` etc would make more sense

File: src/main/java/tools/jackson/databind/ObjectReader.java [hunk 6]
```diff
@@ -1320,6 +1375,187 @@ public <T> T readValue(TokenBuffer src) throws JacksonException
...
+
+                throw new tools.jackson.databind.exc.DeferredBindingException(p, bucket, limitReached);
+            }
+
+            return result;
+
+        } catch (tools.jackson.databind.exc.DeferredBindingException e) {
+            throw e; // Already properly formatted
+
+        } catch (DatabindException e) {
```
Comment (cowtowncoder) [inline src/main/java/tools/jackson/databind/ObjectReader.java]: Is the idea to only catch `DatabindException`s and not other `JacksonException`s?

File: src/main/java/tools/jackson/databind/deser/CollectingProblemHandler.java [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+     * Default maximum number of problems to collect before stopping.
+     * Prevents memory exhaustion attacks.
+     */
+    private static final int DEFAULT_MAX_PROBLEMS = 100;
+
+    /**
+     * Unique private key object for the maximum problem limit attribute.
+     * Using a dedicated object prevents collisions with user attributes.
+     */
+    private static final class MaxProblemsKey {
```
Comment (cowtowncoder) [inline src/main/java/tools/jackson/databind/deser/CollectingProblemHandler.java]: Why would we store configuration as context attribute, instead of within `ProblemHandler` itself, as config setting?

Comment (cowtowncoder) [reply src/main/java/tools/jackson/databind/deser/CollectingProblemHandler.java]: .... or maybe as typed collector state added to context, I guess.

File: src/main/java/tools/jackson/databind/ObjectReader.java [hunk 8]
```diff
@@ -692,6 +692,61 @@ public ObjectReader withHandler(DeserializationProblemHandler h) {
...
+     * @since 3.1
+     */
+    public ObjectReader collectErrors() {
+        return collectErrors(100); // Default limit
+    }
+
+    /**
+     * Enables error collection mode with a custom problem limit.
+     *
+     * <p><b>Thread-safety</b>: The returned reader is immutable and thread-safe.
```
Comment (cowtowncoder) [inline src/main/java/tools/jackson/databind/ObjectReader.java]: Needs to explain that this is implemented as a problem handler, replacing any handler that might have been formerly configured.

File: src/main/java/tools/jackson/databind/ObjectReader.java (hunk 1 above)
Comment (sri-adarsh-kumar) [reply src/main/java/tools/jackson/databind/ObjectReader.java:740]: Added now. Thanks for the feedback.

File: src/main/java/tools/jackson/databind/ObjectReader.java (hunk 2 above)
Comment (sri-adarsh-kumar) [reply src/main/java/tools/jackson/databind/ObjectReader.java]: Changed now. Thanks.

File: src/main/java/tools/jackson/databind/ObjectReader.java (hunk 4 above)
Comment (sri-adarsh-kumar) [reply src/main/java/tools/jackson/databind/ObjectReader.java]: Changed to `problemCollectingReader` and `readValueCollectingProblems` method names

File: src/main/java/tools/jackson/databind/deser/CollectingProblemHandler.java (hunk 5 above)
Comment (sri-adarsh-kumar) [reply src/main/java/tools/jackson/databind/deser/CollectingProblemHandler.java]: Thanks for the pointer 😉 . Changed now.

File: src/main/java/tools/jackson/databind/ObjectReader.java (hunk 8 above)
Comment (sri-adarsh-kumar) [reply src/main/java/tools/jackson/databind/ObjectReader.java]: Added in the Javadoc for `problemCollectingReader` method.

File: src/main/java/tools/jackson/databind/ObjectReader.java (hunk 6 above)
Comment (sri-adarsh-kumar) [reply src/main/java/tools/jackson/databind/ObjectReader.java]: Yes, that is intentional. DatabindException represents deserialization/binding errors that we want to collect (missing properties, type mismatches, etc.). 

Other Jackson exceptions (like JsonParseException, StreamReadException) might pertain more to malformed JSON structure that would fail fast and not be collected.

**Comment by sri-adarsh-kumar** [comment]
@JooHyukKim 

Thanks for the feedback.
> -1 on this PR itself due to the sheer amount of line changes.


If it's some consolation, the actual logic change in ObjectReader and CollectingProblemHandler is only like 1/3 of the total changes. Others are from domain classes, documentation and tests. As this is not a small change (atleast from my POV) this seemed reasonable for me.

> Ai code assistants should really get trained on splitting tasks to safer(or at least feeling so) bits of progression

Reg Coding Assistants and splitting tasks, this is indeed true. Once their context window starts filling up their answers are not that reliable.

Let me explain my process how I overcome this (something I also do in my job).

1. Specify the requirements to Claude Sonnet 4.5 (in Claude Code) along with any context (in this case the Github issue and the underlying discussion) and ask it to make a technical spec.

2. Review the spec with GPT-5 and continuously iterate on it between Sonnet 4.5, GPT-5 and myself. Spec will itself have sub-sections and checklist that the Coding Agent can build piece by piece.

3. Build the finalised spec using the Coding Agent (Sonnet 4.5).

4. Self review manually as well as with GPT-5 until the solution is in a reviewable state

5. Push the solution for peer review 

If you see any code blocks that could be improved/simplified or if you have any feedback to my development process (mentioned above), I would be happy to learn and improve on it in future.

File: src/main/java/tools/jackson/databind/deser/CollectingProblemHandler.java (hunk 7 above)
Comment (sri-adarsh-kumar) [reply src/main/java/tools/jackson/databind/deser/CollectingProblemHandler.java]: @cowtowncoder Can you clarify if there is any change expected for this?
From my understanding, the configuration is currently part of the context as a typed collector state.

**Comment by sri-adarsh-kumar** [comment]
@cowtowncoder Just wanted to check in as the PR hasn't been reviewed yet.
Please let me know if there is any input needed from my side.

**Comment by cowtowncoder** [comment]
@sri-adarsh-kumar I haven't had time to re-review, but not waiting for anything from your side. Biggest question is that of how API should work, functionality be exposed.

File: src/main/java/tools/jackson/databind/ObjectReader.java [hunk 9]
```diff
@@ -1320,6 +1385,207 @@ public <T> T readValue(TokenBuffer src) throws JacksonException
...
+    /**
+     * Convenience overload for {@link #readValueCollectingProblems(JsonParser)}.
+     */
+    public <T> T readValueCollectingProblems(String content) throws JacksonException {
+        _assertNotNull("content", content);
+        DeserializationContextExt ctxt = _deserializationContext();
+        JsonParser p = _considerFilter(_parserFactory.createParser(ctxt, content), true);
+        try {
+            return readValueCollectingProblems(p);
+        } finally {
```
Comment (cowtowncoder) [inline src/main/java/tools/jackson/databind/ObjectReader.java]: Why not use `try (JsonParser p = ....) { ... }` for auto-closing?

File: src/main/java/tools/jackson/databind/deser/CollectingProblemHandler.java [hunk 10]
```diff
@@ -0,0 +1,367 @@
...
+        return true;
+    }
+
+    /**
+     * Safely retrieves the current token location, handling null parser.
+     */
+    private TokenStreamLocation safeGetLocation(JsonParser p) {
+        try {
+            return (p != null) ? p.currentTokenLocation() : null;
+        } catch (Exception e) {
```
Comment (cowtowncoder) [inline src/main/java/tools/jackson/databind/deser/CollectingProblemHandler.java]: No defensive coding like this please; this would quietly eat NPEs and such.

File: src/main/java/tools/jackson/databind/deser/CollectingProblemHandler.java [hunk 11]
```diff
@@ -0,0 +1,367 @@
...
+        }
+    }
+
+    /**
+     * Safely retrieves the current token, handling null parser.
+     */
+    private JsonToken safeGetToken(JsonParser p) {
+        try {
+            return (p != null) ? p.currentToken() : null;
+        } catch (Exception e) {
```
Comment (cowtowncoder) [inline src/main/java/tools/jackson/databind/deser/CollectingProblemHandler.java]: Same here, no.

File: src/main/java/tools/jackson/databind/deser/CollectingProblemHandler.java [hunk 12]
```diff
@@ -0,0 +1,367 @@
...
+     */
+    private JsonPointer buildJsonPointer(JsonParser p) {
+        if (p == null) {
+            return JsonPointer.empty();
+        }
+
+        TokenStreamContext ctx = p.streamReadContext();
+
+        // Collect segments from current to root
+        List<Object> segments = new ArrayList<>(); // String (property) or Integer (index)
```
Comment (cowtowncoder) [inline src/main/java/tools/jackson/databind/deser/CollectingProblemHandler.java]: Why all this code instead of using:

    public JsonPointer pathAsPointer() {

of `TokenStreamContext`?

File: src/main/java/tools/jackson/databind/deser/CollectingProblemHandler.java (hunk 7 above)
Comment (cowtowncoder) [reply src/main/java/tools/jackson/databind/deser/CollectingProblemHandler.java]: Ok, so, storage of accumulated problems needs to be as Attribute, I understand that reading through whole PR.

But "max problems" should simply be config property of `CollectingProblemHandler`, no need to complicate it by externalizing.
So can remove `MaxProblemsKey` I think.

File: src/main/java/tools/jackson/databind/deser/CollectingProblemHandler.java [hunk 13]
```diff
@@ -0,0 +1,367 @@
...
+        );
+
+        // Store null as rawValue for unknown properties
+        // (property name is in the path, no need to duplicate)
+        if (recordProblem(ctxt, message, null, null)) {
+            p.skipChildren(); // Skip the unknown property value
+            return true; // Problem handled
+        }
+
+        return false; // Limit reached, let default handling throw
```
Comment (cowtowncoder) [inline src/main/java/tools/jackson/databind/deser/CollectingProblemHandler.java]: Should we not instead throw "full" `DeferredBindingException`? To avoid losing 100 (etc) collected problems?

File: src/main/java/tools/jackson/databind/deser/CollectingProblemHandler.java [hunk 14]
```diff
@@ -0,0 +1,367 @@
...
+    }
+
+    /**
+     * Returns a sensible default value for the given type to allow
+     * deserialization to continue.
+     *
+     * <p>IMPORTANT: Only primitives get non-null defaults. Reference types
+     * (including boxed primitives) get null to avoid masking nullability issues.
+     */
+    private Object getDefaultValue(Class<?> type) {
```
Comment (cowtowncoder) [inline src/main/java/tools/jackson/databind/deser/CollectingProblemHandler.java:305]: Use existing `ClassUtil.defaultValue(type)`.

File: src/main/java/tools/jackson/databind/deser/CollectingProblemHandler.java [hunk 15]
```diff
@@ -0,0 +1,367 @@
...
+        // This avoids masking nullability issues in the domain model
+        return null;
+    }
+
+    /**
+     * Checks if it's safe to return null for a given type after
+     * instantiation failure.
+     */
+    private boolean canReturnNullFor(Class<?> type) {
+        // Cannot return null for primitives or arrays
```
Comment (cowtowncoder) [inline src/main/java/tools/jackson/databind/deser/CollectingProblemHandler.java]: Why not for Arrays? `null` is valid for them, similar to `Collection`s etc (and unlike primitives)

File: src/main/java/tools/jackson/databind/ObjectReader.java [hunk 16]
```diff
@@ -692,6 +695,68 @@ public ObjectReader withHandler(DeserializationProblemHandler h) {
...
+     * @since 3.1
+     */
+    public ObjectReader problemCollectingReader(int maxProblems) {
+        if (maxProblems <= 0) {
+            throw new IllegalArgumentException("maxProblems must be positive");
+        }
+
+        // Store ONLY the max limit in config (not the bucket)
+        // Bucket will be allocated fresh per-call in readValueCollectingProblems()
+        ContextAttributes attrs = _config.getAttributes()
```
Comment (cowtowncoder) [inline src/main/java/tools/jackson/databind/ObjectReader.java]: As per my other notes, no need for this: max-count can (and should) be part of of collector itself.
Can thus avoid adding shared attribute(s).

File: src/main/java/tools/jackson/databind/ObjectReader.java [hunk 17]
```diff
@@ -1320,6 +1385,207 @@ public <T> T readValue(TokenBuffer src) throws JacksonException
...
+                throw new DeferredBindingException(p, bucket, limitReached);
+            }
+
+            return result;
+
+        } catch (DeferredBindingException e) {
+            throw e; // Already properly formatted
+
+        } catch (DatabindException e) {
+            // Hard failure occurred; attach collected problems as suppressed
```
Comment (cowtowncoder) [inline src/main/java/tools/jackson/databind/ObjectReader.java]: Good idea!

File: src/main/java/tools/jackson/databind/ObjectReader.java [hunk 18]
```diff
@@ -1320,6 +1385,207 @@ public <T> T readValue(TokenBuffer src) throws JacksonException
...
+     * parser filtering. Callers are responsible for filter wrapping if needed.
+     *
+     * @param <T> Type to deserialize
+     * @param p JsonParser to read from (will not be closed by this method)
+     * @return Deserialized object
+     * @throws DeferredBindingException if recoverable problems were collected
+     * @throws DatabindException if a non-recoverable error occurred
+     * @since 3.1
+     */
+    public <T> T readValueCollectingProblems(JsonParser p) throws JacksonException {
```
Comment (cowtowncoder) [inline src/main/java/tools/jackson/databind/ObjectReader.java:1453]: Ok: I think we should re-factor this, to avoid construction of multiple `DeserializationContextExt` instances -- see how existing `readValue(String)` (f.ex) and `readValue(JsonParser)` differ (they call different methods (`_bind()` vs `_bindAndClose()`).

But more importantly, we should not have to create new `ObjectReader` here.

File: src/main/java/tools/jackson/databind/ObjectReader.java [hunk 19]
```diff
@@ -1320,6 +1385,207 @@ public <T> T readValue(TokenBuffer src) throws JacksonException
...
+        _assertNotNull("p", p);
+
+        // CRITICAL: Allocate a FRESH bucket for THIS call (thread-safety)
+        List<CollectedProblem> bucket = new ArrayList<>();
+
+        // Create per-call attributes with the fresh bucket
+        ContextAttributes perCallAttrs = _config.getAttributes()
```
Comment (cowtowncoder) [inline src/main/java/tools/jackson/databind/ObjectReader.java]: Instead of this, let's make use of `DeserializationContext.setAttribute()` -- `DeserializationContext` is per-call, mutable. 

And to have access to the DC, need to refactor to have something like

    protected <T> T readValueCollectingProblems(JsonParser p, DeserializationContextExt ctxt) 

that all public `readValueCollectingProblems()` call with parser, context created.

Finally, we should NOT call `readValue(parser)` because that would just create yet another `DeserializationContext`. This means it'll be necessary to call one `_bind()` or `_bindAndClose()`.

File: src/main/java/tools/jackson/databind/ObjectReader.java (hunk 9 above)
Comment (cowtowncoder) [reply src/main/java/tools/jackson/databind/ObjectReader.java]: ... I can actually fix these.

File: src/main/java/tools/jackson/databind/deser/CollectingProblemHandler.java [hunk 20]
```diff
@@ -0,0 +1,367 @@
...
+
+    @Override
+    public boolean handleUnknownProperty(DeserializationContext ctxt,
+            JsonParser p, ValueDeserializer<?> deserializer,
+            Object beanOrClass, String propertyName) throws JacksonException {
+
+        String message = String.format(
+            "Unknown property '%s' for type %s",
+            propertyName,
+            beanOrClass instanceof Class ?
```
Comment (cowtowncoder) [inline src/main/java/tools/jackson/databind/deser/CollectingProblemHandler.java]: Use `ClassUtil.getClassDescription(Object classOrInstance)`

**Comment by cowtowncoder** [comment]
Ok. Phew! Had another pass and now I understand this much better. I like the fact this one of remaining "Most Wanted" issues. There are a few things to address so added quite a few notes. But we are getting there. :)

(and once again, I am EXCITED about solving this long-standing HIGHLY VOTED FOR feature request!)

File: src/main/java/tools/jackson/databind/deser/CollectingProblemHandler.java (hunk 7 above)
Comment (sri-adarsh-kumar) [reply src/main/java/tools/jackson/databind/deser/CollectingProblemHandler.java]: Changed as per suggestion.

File: src/main/java/tools/jackson/databind/deser/CollectingProblemHandler.java (hunk 10 above)
Comment (sri-adarsh-kumar) [reply src/main/java/tools/jackson/databind/deser/CollectingProblemHandler.java]: Removed.

File: src/main/java/tools/jackson/databind/deser/CollectingProblemHandler.java (hunk 11 above)
Comment (sri-adarsh-kumar) [reply src/main/java/tools/jackson/databind/deser/CollectingProblemHandler.java]: Removed.

File: src/main/java/tools/jackson/databind/deser/CollectingProblemHandler.java (hunk 12 above)
Comment (sri-adarsh-kumar) [reply src/main/java/tools/jackson/databind/deser/CollectingProblemHandler.java]: Changed as per suggestion.

File: src/main/java/tools/jackson/databind/deser/CollectingProblemHandler.java (hunk 14 above)
Comment (sri-adarsh-kumar) [reply src/main/java/tools/jackson/databind/deser/CollectingProblemHandler.java:305]: Changed as per suggestion.

File: src/main/java/tools/jackson/databind/deser/CollectingProblemHandler.java (hunk 15 above)
Comment (sri-adarsh-kumar) [reply src/main/java/tools/jackson/databind/deser/CollectingProblemHandler.java]: True. Changed to just check for primitives.

File: src/main/java/tools/jackson/databind/deser/CollectingProblemHandler.java (hunk 13 above)
Comment (sri-adarsh-kumar) [reply src/main/java/tools/jackson/databind/deser/CollectingProblemHandler.java]: This is how it should be. This is also verified under `CollectingErrorsTest#defaultLimit()` test

File: src/main/java/tools/jackson/databind/ObjectReader.java (hunk 16 above)
Comment (sri-adarsh-kumar) [reply src/main/java/tools/jackson/databind/ObjectReader.java]: Changed as per suggestion.

File: src/main/java/tools/jackson/databind/deser/CollectingProblemHandler.java (hunk 20 above)
Comment (sri-adarsh-kumar) [reply src/main/java/tools/jackson/databind/deser/CollectingProblemHandler.java]: Changed as per suggestion.

File: src/main/java/tools/jackson/databind/ObjectReader.java (hunk 18 above)
Comment (sri-adarsh-kumar) [reply src/main/java/tools/jackson/databind/ObjectReader.java:1453]: `bind` and `bindAndClose` have been separated now in context of problem collection.
There is now `_collectingBind` and `_collectingBindAndClose` following a similar distinction

File: src/main/java/tools/jackson/databind/ObjectReader.java (hunk 19 above)
Comment (sri-adarsh-kumar) [reply src/main/java/tools/jackson/databind/ObjectReader.java]: Refactored `readValueCollectingProblems` so every overload shares a single `DeserializationContextExt`, stores the bucket via `ctxt.setAttribute(...)`, and calls `_bind`/`_collectingBindAndClose`  directly without creating new ObjectReader instances.

File: src/main/java/tools/jackson/databind/deser/CollectingProblemHandler.java (hunk 13 above)
Comment (cowtowncoder) [reply src/main/java/tools/jackson/databind/deser/CollectingProblemHandler.java]: Ok.

**Comment by cowtowncoder** [comment]
@sri-adarsh-kumar Ok I am reviewing this again and things look good; hoping to get this approved and merged soon.

But one process thing: do we have a CLA (or possibly CCLA, from... Zalando?) for you? If so, LMK and I'll re-check.
If not, we would need one before merging (but that one is good for all future contributions):

https://github.com/FasterXML/jackson/blob/main/contributor-agreement.pdf

and the usual way is to print it, fill & sign, scan/photo, email to `cla` at fasterxml dot com.
Once I have it (and like I said, if you have sent one before and I can verify that's good), I can proceed with merging.

Looking forward to getting this cool feature added in 3.1!

**Comment by sri-adarsh-kumar** [comment]
Thanks for the review.

There should be a CCLA from Zalando as part of 
https://github.com/FasterXML/jackson-datatypes-misc/pull/48

This was also referenced in https://github.com/FasterXML/jackson-datatypes-misc/pull/76

**Comment by cowtowncoder** [comment]
@sri-adarsh-kumar Ok thanks: I did look at CCLA but did not realize it just referred to:

 https://github.com/orgs/zalando/people

for list of employees and matching accounts (which is fine). So that should be good. Thank you for clarifying.
